File: backend/src/webex.py
import asyncio
import details
import scribe
from playwright.async_api import TimeoutError
from datetime import datetime


async def meeting(page):

    print("Getting meeting link.")
    await page.goto("https://signin.webex.com/join")

    print("Entering meeting ID.")
    meeting_text_element = await page.wait_for_selector("#join-meeting-form")
    await meeting_text_element.type(details.meeting_id)
    await meeting_text_element.press("Enter")

    print("Launching app.")
    await page.wait_for_selector(".meet_message_H1")
    await page.goto(f"{page.url}?launchApp=true")

    frame_element = await page.wait_for_selector("iframe[name='thinIframe']")
    frame = await frame_element.content_frame()

    print("Entering name.")
    name_text_element = await frame.wait_for_selector(
        'input[aria-labelledby="nameLabel"]'
    )
    await name_text_element.type(details.scribe_identity)

    print("Entering email.")
    email_text_element = await frame.wait_for_selector(
        'input[aria-labelledby="emailLabel"]'
    )
    # await email_text_element.type(details.email_sender)
    await email_text_element.type("scribe@scribe.dev.amazon.com")
    await email_text_element.press("Enter")

    print("Clicking cookie button.")
    cookie_button_element = await page.wait_for_selector(".cookie-manage-close-handler")
    await cookie_button_element.click()

    print("Clicking mute button.")
    mute_button_element = await frame.wait_for_selector('text="Mute"')
    await mute_button_element.click()

    print("Clicking video button.")
    video_button_element = await frame.wait_for_selector('text="Stop video"')
    await video_button_element.click()

    print("Clicking join button.")
    join_button_element = await frame.wait_for_selector('text="Join meeting"')
    await join_button_element.click()

    print("Opening chat panel.")
    try:
        chat_panel_element = await frame.wait_for_selector(
            'text="Chat"',
            timeout=details.waiting_timeout,
        )
    except TimeoutError:
        print("Your scribe was not admitted into the meeting.")
        return
    else:
        await chat_panel_element.click()

    async def send_messages(messages):
        message_element = await frame.wait_for_selector(
            'textarea[placeholder="Type your message here"]'
        )
        for message in messages:
            await message_element.type(message)
            await message_element.press("Enter")

    print("Sending introduction messages.")
    await send_messages(details.intro_messages)

    # async def attendee_change(number: int):
    #     if number <= 1:
    #         print("Your scribe got lonely and left.")
    #         await page.goto("about:blank")

    # await page.expose_function("attendeeChange", attendee_change)

    # print("Listening for attendee changes.")
    # await page.evaluate(
    #     """
    #     const targetNode = document.querySelector('button[data-testid="collapse-container"][aria-label^="Present"]')
    #     const config = { characterData: true, subtree: true }

    #     const callback = (mutationList, observer) => {
    #         attendeeChange(parseInt(mutationList[mutationList.length - 1].target.textContent))
    #     }

    #     const observer = new MutationObserver(callback)
    #     observer.observe(targetNode, config)
    # """
    # )

    await page.expose_function("speakerChange", scribe.speaker_change)

    print("Listening for speaker changes.")
    await page.evaluate(
        """
        const iFrame = document.querySelector("iframe[name='thinIframe']")
        const iFrameDocument = iFrame.contentDocument
        const targetNode = iFrameDocument.querySelector('div[class*="layout-layout-content-left"]')

        const config = { attributes: true, subtree: true };

        const callback = (mutationList, observer) => {
            for (const mutation of mutationList) {
                if (mutation.attributeName === 'class') {
                    const childNode = mutation.target;
                    const pattern = /.*videoitem-in-speaking.*/;
                    if (childNode.classList.value.match(pattern)) {
                        speakerChange(childNode.textContent);
                    }
                }
            }
        }

        const observer = new MutationObserver(callback)
        observer.observe(targetNode, config)
    """
    )

    # async def message_change(sender, text, attachment_title, attachment_href):
    #     global prev_sender
    #     if not sender:
    #         sender = prev_sender
    #     prev_sender = sender
    #     if text == details.end_command:
    #         print("Your scribe has been removed from the meeting.")
    #         await page.goto("about:blank")
    #     elif details.start and text == details.pause_command:
    #         details.start = False
    #         print(details.pause_messages[0])
    #         await send_messages(details.pause_messages)
    #     elif not details.start and text == details.start_command:
    #         details.start = True
    #         print(details.start_messages[0])
    #         await send_messages(details.start_messages)
    #         asyncio.create_task(scribe.transcribe())
    #     elif details.start and not (
    #         sender == "Amazon Chime" or details.scribe_name in sender
    #     ):
    #         timestamp = datetime.now().strftime("%H:%M")
    #         message = f"[{timestamp}] {sender}: "
    #         if attachment_title and attachment_href:
    #             details.attachments[attachment_title] = attachment_href
    #             if text:
    #                 message += f"{text} | {attachment_title}"
    #             else:
    #                 message += attachment_title
    #         else:
    #             message += text
    #         # print('New Message:', message)
    #         details.messages.append(message)

    # await page.expose_function("messageChange", message_change)

    # print("Listening for message changes.")
    # await page.evaluate(
    #     """
    #     const targetNode = document.querySelector('._2B9DdDvc2PdUbvEGXfOU20')
    #     const config = { childList: true, subtree: true }

    #     const callback = (mutationList, observer) => {
    #         for (const mutation of mutationList) {
    #             const addedNode = mutation.addedNodes[0]
    #             if (addedNode) {
    #                 const sender = addedNode.querySelector('h3[data-testid="chat-bubble-sender-name"]')?.textContent
    #                 const text = addedNode.querySelector('.Linkify')?.textContent
    #                 const attachmentElement = addedNode.querySelector('.SLFfm3Dwo5MfFzks4uM11')
    #                 const attachmentTitle = attachmentElement?.title
    #                 const attachmentHref = attachmentElement?.href
    #                 messageChange(sender, text, attachmentTitle, attachmentHref)
    #             }
    #         }
    #     }

    #     const observer = new MutationObserver(callback)
    #     observer.observe(targetNode, config)
    # """
    # )

    print("Waiting for meeting end.")
    try:
        await frame.wait_for_selector(
            ".style-end-message-2PkYs", timeout=details.meeting_timeout
        )
        print("Meeting ended.")
    except TimeoutError:
        print("Meeting timed out.")
    finally:
        details.start = False

File: src/backend/task/src/webex.ts

import { Page, Frame } from 'playwright';
import { transcriptionService } from './scribe';
import { details } from './details';

export default class Webex {
    private readonly iframe = 'iframe[name="thinIframe"]';

    private async sendMessages(frame: Frame, messages: string[]): Promise<void> {
        const messageElement = await frame.waitForSelector(
            'textarea[placeholder="Type your message here"]'
        );
        for (const message of messages) {
            await messageElement?.type(message);
            await messageElement?.press("Enter");
        }
    }

    public async initialize(page: Page): Promise<void> {
        console.log("Getting meeting link.");
        await page.goto("https://signin.webex.com/join");

        console.log("Entering meeting ID.");
        const meetingTextElement = await page.waitForSelector("#join-meeting-form");
        await meetingTextElement?.type(details.meeting_id);
        await meetingTextElement?.press("Enter");

        console.log("Launching app.");
        try {
            await page.waitForSelector(".meet_message_H1");
            await page.goto(`${page.url()}?launchApp=true`);
        } catch {
            console.log("Your scribe was unable to join the meeting.");
            return;
        }

        const frameElement = await page.waitForSelector(this.iframe);
        const frame = await frameElement?.contentFrame();
        if (!frame) return;

        console.log("Entering name.");
        const nameTextElement = await frame.waitForSelector(
            'input[aria-labelledby="nameLabel"]'
        );
        await nameTextElement?.type(details.scribe_identity);

        console.log("Entering email.");
        const emailTextElement = await frame.waitForSelector(
            'input[aria-labelledby="emailLabel"]'
        );
        await emailTextElement?.type(process.env.EMAIL_SOURCE!);
        await emailTextElement?.press("Enter");

        console.log("Clicking cookie button.");
        const cookieButtonElement = await page.waitForSelector(".cookie-manage-close-handler");
        await cookieButtonElement?.click();

        console.log("Clicking mute button.");
        const muteButtonElement = await frame.waitForSelector('text="Mute"');
        await muteButtonElement?.click();

        console.log("Clicking video button.");
        const videoButtonElement = await frame.waitForSelector('text="Stop video"');
        await videoButtonElement?.click();

        console.log("Clicking join button.");
        const joinButtonElement = await frame.waitForSelector('text="Join meeting"');
        await joinButtonElement?.click();

        console.log("Opening chat panel.");
        try {
            const chatPanelElement = await frame.waitForSelector(
                'text="Chat"',
                { timeout: details.waiting_timeout }
            );
            await chatPanelElement?.click();
        } catch {
            console.log("Your scribe was not admitted into the meeting.");
            return;
        }

        await new Promise(resolve => setTimeout(resolve, 1000));

        console.log("Sending introduction messages.");
        await this.sendMessages(frame, details.intro_messages);

        await page.exposeFunction("speakerChange", transcriptionService.speakerChange);
        console.log("Listening for speaker changes.")
        await page.evaluate(
            ({ iframe }) => {
                const iFrame = document.querySelector(iframe);
                const iFrameDocument = iFrame?.ownerDocument;
                const targetNode = iFrameDocument?.querySelector('div[class*="layout-layout-content-left"]');

                const config = { attributes: true, subtree: true };

                const callback = (mutationList: MutationRecord[]) => {
                    for (const mutation of mutationList) {
                        if (mutation.attributeName === 'class') {
                            const childNode = mutation.target as HTMLElement;
                            const pattern = /.*videoitem-in-speaking.*/;
                            if (childNode.classList.value.match(pattern)) {
                                (window as any).speakerChange(childNode.textContent);
                            }
                        }
                    }
                };

                const observer = new MutationObserver(callback);
                if (targetNode) observer.observe(targetNode, config);
            },
            { iframe: this.iframe }
        );

        await page.exposeFunction("messageChange", async (sender: string, message: string) => {
            if (!sender?.startsWith("Scribe")) {
                if (message.includes(details.end_command)) {
                    console.log("Your scribe has been removed from the meeting.");
                    await page.goto("about:blank");
                } else if (details.start && message.includes(details.pause_command)) {
                    details.start = false;
                    console.log(details.pause_messages[0]);
                    await this.sendMessages(frame, details.pause_messages);
                } else if (!details.start && message.includes(details.start_command)) {
                    details.start = true;
                    console.log(details.start_messages[0]);
                    await this.sendMessages(frame, details.start_messages);
                } else if (details.start) {
                    details.messages.push(message);
                }
            }
        });
        console.log("Listening for message changes.")
        await page.evaluate(
            ({ iframe }) => {
                const iFrame = document.querySelector(iframe);
                const iFrameDocument = iFrame?.ownerDocument;
                const targetNode = iFrameDocument?.querySelector('div[class^="style-chat-box"]');
                const config = { childList: true, subtree: true };

                const callback = (mutationList: MutationRecord[]) => {
                    const lastMutation = mutationList[mutationList.length - 1];
                    const addedNode = lastMutation.addedNodes[0] as Element;
                    if (addedNode) {
                        const sender = addedNode.querySelector('h3[class^="style-chat-label"]')?.textContent;
                        const message = addedNode.querySelector('span[class^="style-chat-msg"]')?.textContent;
                        if (sender && message) {
                            (window as any).messageChange(sender, message);
                        }
                    }
                };

                const observer = new MutationObserver(callback);
                if (targetNode) observer.observe(targetNode, config);
            },
            { iframe: this.iframe }
        );

        console.log("Waiting for meeting end.");
        try {
            await page.waitForSelector('.style-end-message-2PkYs', {
                timeout: details.meeting_timeout
            });
            console.log("Meeting ended.");
        } catch (error) {
            console.log("Meeting timed out.");
        } finally {
            details.start = false;
        }
    }

}
